openid-4-verifiable-presentations-1_0.md

```diff
@@ -243,11 +243,26 @@ when it receives `request_uri_method` parameter with the value `post` but does n
+## Client Identifier Scheme
+
+This specification defines the concept of a Client Identifier Scheme that indicates how the Wallet is supposed to interpret the Client Identifier and associated data in the process of Client identification, authentication, and authorization. It is communicated in prefix within the `client_id` parameter. This prefix enables deployments of this specification to use different mechanisms to obtain and validate Client metadata beyond the scope of [@!RFC6749]. A certain Client Identifier Scheme MAY require the Verifier to sign the Authorization Request as means of authentication and/or pass additional parameters and require the Wallet to process them.
+
+The value of the `client_id` Authorization Request parameter, as defined in [@!RFC6749], therefore MUST be structured as follows: 
+```
+<client_id_scheme>:<orig_client_id>
+```
+
+Here, `<client_id_scheme>` is the Client Identifier Scheme and `<orig_client_id>` is the identifier for the Client within the namespace of that scheme. See (#client_metadata_management) for Client Identifier Schemes defined by this specification. 
+
+For example, a request might contain `client_id=verifier_attestation:example-client` to indicate that the `verifier_attestation` Client Identifier Scheme is to be used and that within this scheme, the Verifier can be identified by the string `example-client`. The presentation would contain the string `verifier_attestation:example-client` as the audience (intended receiver). In all other places within the OAuth flow where the Client Identifier is used, the full string `verifier_attestation:example-client` would be used.
+
+Important: Without the prefix, the `orig_client_id` is not sufficient to identify the Verifier, as different Verifiers (under control by different entities and using different cryptographic material) might have the same `orig_client_id` within different `client_id_scheme` namespaces. Therefore, `orig_client_id` MUST NOT be used independently within the context of the Wallet or its responses to identify the client. Only the full string `<client_id_scheme>:<orig_client_id>` is the Client Identifier of the Verifier, in particular in places where the Client Identifier is used in [@!RFC6749] and in the presentation returned to the Verifier.
+
+Note that the Verifier needs to determine which Client Identifier schemes the Wallet supports prior to sending the Authorization Request in order to choose a supported scheme.
```

Comment:
As I've highlighted before this is a huge limitation of this approach and is the basis of discussion in #248 and #247. It essentially means that there needs be some form of out of band negotiation model in order for a verifier to pick the right scheme. I don't see how this is practically possible.

Reply:
The changes in https://github.com/openid/OpenID4VP/issues/248 allow this to work and can be easily updated to work with the changes in this PR (i.e. putting everything in `client_id` instead of having a separate client_id_scheme in the unprotected header). I can't see any real practical issues with that approach.

we only have two proposals on how to fix this on the table and (as I said on the issue) I don't think the suggestion in #247 fully solves the problem.

Reply:
Thanks, I've left a comment on #248.

Reply:
Resolving this as the discussion is continued in #248.